File: solutions/fsharp/log-levels/1/LogLevels.fs
module LogLevels

let level = function
  | "[WARNING]" -> "warning"
  | "[ERROR]" -> "error"
  | "[INFO]" -> "info"
  | s -> s

let re = function
  | "[WARNING]" -> "(warning)"
  | "[ERROR]" -> "(error)"
  | "[INFO]" -> "(info)"
  | s -> s

let format (formatter: string -> string) (logline: string) = logline.Split(':') |> Seq.map ((fun s -> s.Trim()) >> formatter)

let message = format level >> Seq.last

let logLevel = format level >> Seq.head

let reformat = format re >> Seq.rev >> Seq.reduce (fun acc s -> acc + " " + s)

File: solutions/csharp/log-levels/1/LogLevels.cs
using System;
using System.Text.RegularExpressions;

static class LogLine
{   
    private static string GetLevel(string log) => new Regex(@"\[(?<level>\w+)\]").Match(log).Groups["level"].Value.ToLowerInvariant();
    private static string GetMessage(string log) => new Regex(@"\s+(?<message>[\w\s]+)\s*").Match(log).Groups["message"].Value.Trim();

    public static string Message(string logLine) => GetMessage(logLine);
    public static string LogLevel(string logLine) => GetLevel(logLine);
    public static string Reformat(string logLine) => $"{GetMessage(logLine)} ({GetLevel(logLine)})";
}